File: 프로그래머스/1/86491. 최소직사각형/최소직사각형.swift
import Foundation

func solution(_ sizes:[[Int]]) -> Int {
    
    var w: Int = 0
    var h: Int = 0
    
    for array in sizes {
        let sortedArray = array.sorted()
        w = max(w, sortedArray.first ?? 0)
        h = max(h, sortedArray.last ?? 0)
    }
    
    return w * h
}

// 지갑의 크기
// size: 1~1만
// w, h: 1~1000

// 1만개 -> O(N^2)

// 긴거를 기준으로 맞추자. 가로 세로 중
// sizes를 순회하며 긴 거로 Swap
// 각 sizes의 Max 값을 리턴

File: 프로그래머스/1/86491. 최소직사각형/최소직사각형.py
# 지갑의 크기
# 모두 수납 가능하면서 작은 크기
# 모든 지갑을 수납할 수 있는 가장 작은 지갑의 크기

# n: 1~1만
# 회전이 가능하다.
# 큰, 작은으로 각각 max
# O(1만)

def solution(sizes):
    
    max_w = -1
    max_h = -1
    
    # 항상 가로가 긴 지갑 생성
    for size in sizes:
        max_w = max(max_w, max(size))
        max_h = max(max_h, min(size))
    
    return max_w * max_h

# 3:38초